File: server/lib/sweep.ts
/**
 * Hyperparameter sweep search-space expansion.
 *
 * Deliberately not Ludwig's own `hyperopt` (Ray Tune-backed): that would
 * contend with training for the single GPU worker, its progress is opaque
 * to the run/metric/log event pipeline this app already has, and it reuses
 * none of the cancel/DLQ/reaper machinery `queueTraining` already gives a
 * manually-started run. Instead a sweep is just N ordinary training runs —
 * this module's only job is turning a search space into that list of N
 * `TrainerSelections`, which `routes/sweeps.ts` then dispatches one at a
 * time through the existing `queueTraining` (server/lib/microservice.ts).
 */

import type { TrainerSelections } from '@server/lib/ludwig'

/** One trainer knob → its candidate values for this sweep. Keys mirror TrainerSelections. */
export type SweepSearchSpace = Partial<Record<keyof TrainerSelections, (number | string)[]>>

const MAX_TRIALS_CAP = 50

export function validateSearchSpace(searchSpace: SweepSearchSpace, maxTrials: number): string | null {
  const keys = Object.keys(searchSpace)
  if (keys.length === 0) return 'Search space must specify at least one hyperparameter'
  for (const key of keys) {
    const values = searchSpace[key as keyof SweepSearchSpace]
    if (!values || values.length === 0) return `'${key}' must list at least one candidate value`
  }
  if (maxTrials < 1 || maxTrials > MAX_TRIALS_CAP) return `maxTrials must be between 1 and ${MAX_TRIALS_CAP}`
  return null
}

/** Every combination of the search space's candidate values (cartesian product), in a stable order. */
export function expandGrid(searchSpace: SweepSearchSpace): TrainerSelections[] {
  const entries = Object.entries(searchSpace) as [keyof TrainerSelections, (number | string)[]][]
  return entries.reduce<TrainerSelections[]>(
    (combos, [key, values]) => combos.flatMap((combo) => values.map((value) => ({ ...combo, [key]: value }))),
    [{}],
  )
}

/** `count` combinations, each knob sampled independently and uniformly from its candidates. */
export function sampleRandom(searchSpace: SweepSearchSpace, count: number): TrainerSelections[] {
  const entries = Object.entries(searchSpace) as [keyof TrainerSelections, (number | string)[]][]
  return Array.from({ length: count }, () => {
    // Built loosely-typed and cast once at the end, rather than per-key —
    // TrainerSelections' value type varies per key, but the candidate list
    // for each key was declared against that same key in SweepSearchSpace.
    const trial: Record<string, number | string> = {}
    for (const [key, values] of entries) {
      trial[key] = values[Math.floor(Math.random() * values.length)]
    }
    return trial as TrainerSelections
  })
}

/**
 * Expand a search space into the trials to dispatch. `grid` is capped at
 * `maxTrials` (truncated, not sampled down — a user who wants the full grid
 * sets maxTrials to the product size) so a large search space can't
 * accidentally dispatch hundreds of runs against the single GPU worker.
 */
export function expandSweep(
  searchSpace: SweepSearchSpace,
  strategy: 'grid' | 'random',
  maxTrials: number,
): TrainerSelections[] {
  if (strategy === 'grid') return expandGrid(searchSpace).slice(0, maxTrials)
  return sampleRandom(searchSpace, maxTrials)
}

File: ai_service/theseus/services/sweep.py
"""Hyperparameter sweep search-space expansion.

Deliberately not Ludwig's own hyperopt (Ray Tune-backed): that would contend with training for
the single GPU, its progress is opaque to the run/metric/log event pipeline, and it reuses none
of the cancel and recovery machinery an ordinary run already gets. Instead a sweep is just N
ordinary training runs: this module turns a search space into that list of N selections, which
the sweeps router then enqueues one after another through the normal training path.

Keys are the camelCase names of TrainerSelections (learningRate, batchSize, encoderId, ...),
exactly as they arrive in the API and are stored in sweeps.search_space.
"""

import itertools
import random
from typing import Any, Literal

MAX_TRIALS_CAP = 50

SearchSpace = dict[str, list[Any]]


def validate_search_space(search_space: SearchSpace, max_trials: int) -> str | None:
    """Returns an error message, or None when the search space is valid."""
    if not search_space:
        return "Search space must specify at least one hyperparameter"
    for key, values in search_space.items():
        if not values:
            return f"'{key}' must list at least one candidate value"
    if max_trials < 1 or max_trials > MAX_TRIALS_CAP:
        return f"maxTrials must be between 1 and {MAX_TRIALS_CAP}"
    return None


def expand_grid(search_space: SearchSpace) -> list[dict[str, Any]]:
    """Every combination of the candidate values (cartesian product), in a stable order.

    The first key varies slowest. An empty search space yields exactly one empty trial.
    """
    keys = list(search_space)
    return [dict(zip(keys, combo, strict=True)) for combo in itertools.product(*(search_space[k] for k in keys))]


def sample_random(search_space: SearchSpace, count: int, rng: random.Random | None = None) -> list[dict[str, Any]]:
    """`count` combinations, each knob sampled independently and uniformly from its candidates."""
    rng = rng or random.Random()
    return [{key: rng.choice(values) for key, values in search_space.items()} for _ in range(count)]


def expand_sweep(
    search_space: SearchSpace, strategy: Literal["grid", "random"], max_trials: int, rng: random.Random | None = None
) -> list[dict[str, Any]]:
    """Expand a search space into the trials to dispatch.

    `grid` is capped at max_trials by truncating, not sampling down (a user who wants the full
    grid sets max_trials to the product size), so a large space cannot accidentally enqueue
    hundreds of runs against the single GPU.
    """
    if strategy == "grid":
        return expand_grid(search_space)[:max_trials]
    return sample_random(search_space, max_trials, rng)
